src/utils/records.ts

```diff
@@ -272,14 +272,54 @@ export class Records {
     return filterCopy;
   }
 
+
+  public static tagsFilterIsPrefix(filter: RecordsWriteTagsFilter): filter is PrefixFilter {
+    return typeof filter === 'object' && ('prefix' in filter && typeof filter.prefix === 'string');
+  }
+
+  /**
+   * This will create individual keys for each of the tags that look like `tag.tag_property`
+   */
+  public static flattenTags(tags: RecordsWriteTags): KeyValues {
+    const tagValues:KeyValues = {};
+    for (const property in tags) {
+      const value = tags[property];
+      tagValues[`tag.${property}`] = value;
+    }
+    return tagValues;
+  }
+
+  /**
+   * This will create individual keys for each of the tag filters that look like `tag.tag_filter_property`
+   */
+  public static flattenTagFilters( tags: { [property: string]: RecordsWriteTagsFilter}): Filter {
+    const tagValues:Filter = {};
+    for (const property in tags) {
+      const value = tags[property];
+
+      tagValues[`tag.${property}`] = this.tagsFilterIsPrefix(value) ? FilterUtility.constructPrefixFilterAsRangeFilter(value.prefix) : value;
+    }
+    return tagValues;
+  }
+
   /**
    *  Converts an incoming RecordsFilter into a Filter usable by MessageStore.
    *
    * @param filter A RecordsFilter
    * @returns {Filter} a generic Filter able to be used with MessageStore.
    */
   public static convertFilter(filter: RecordsFilter, dateSort?: DateSort): Filter {
-    const filterCopy = { ...filter } as Filter;
+    // we process tags separately from the remaining filters.
+    // this is because we prepend each field within the `tags` object with a `tag.` to bring it into the flat-space filter.
```

Comment:
"flat-space" is used a term vs "protocol-space", so let's avoid reusing it.

```suggestion
    // this is because we prepend each field within the `tags` object with a `tag.` to avoid name clashing with first-class index keys.
```